proposals/0257-string-uninitialized-initializer.md

```diff
@@ -0,0 +1,141 @@
+# Add a String Initializer with Access to Uninitialized Storage
+
+* Proposal: [SE-0257](0257-string-uninitialized-initializer.md)
+* Author: [David Smith](https://github.com/Catfish-Man)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: [apple/swift#23409](https://github.com/apple/swift/pull/23409)
+* Bug: [SR-10288](https://bugs.swift.org/browse/SR-10288)
+
+## Introduction
+
+This proposal suggests a new initializer for `String` that provides access to a String's uninitialized storage buffer.
+
+## Motivation
+
+`String` today is well-suited to interoperability with raw memory buffers when a contiguous buffer is already available, such as when dealing with `malloc`ed C strings. However, there are quite a few situations where no such buffer is available, requiring a temporary one to be allocated and copied into. One example is bridging `NSString` to `String`, which currently uses standard library internals to get good performance when using `CFStringGetBytes`. Another, also from the standard library, is `Int` and `Float`, which currently create temporary stack buffers and do extra copying. We expect libraries like SwiftNIO will also find this useful for dealing with streaming data.
+
+## Proposed solution
+
+Add a new `String` initializer that lets a program work with an uninitialized
+buffer.
+
+The new initializer takes a closure that operates on an
+`UnsafeMutableBufferPointer` and an `inout` count of initialized elements. This
+closure has access to the uninitialized contents of the newly created String's
+storage, and must set the intialized count of the array before exiting.
+
+```swift
+let myCocoaString = NSString("The quick brown fox jumps over the lazy dog") as CFString
+var myString = String(unsafeUninitializedCapacity: CFStringGetMaximumSizeForEncoding(myCocoaString, …)) { buffer, initializedCount in
+    CFStringGetBytes(
+    	myCocoaString,
+    	buffer,
+    	…,
+    	&initializedCount
+    )
+}
+// myString == "The quick brown fox jumps over the lazy dog"
+```
+
+Without this initializer we would have had to heap allocate an `UnsafeMutableBufferPointer`, copy the `NSString` contents into it, and then copy the buffer again as we initialized the `String`.
+
+## Detailed design
+
+```swift
+  /// Creates a new String with the specified capacity in UTF-8 code units then
+  /// calls the given closure with a buffer covering the String's uninitialized
+  /// memory.
+  ///
+  /// The closure should set `initializedCount` to the number of
+  /// initialized code units, or 0 if it couldn't initialize the buffer
+  /// (for example if the requested capacity was too small).
+  ///
```

Comment:
Why is setting to `0` if it's too small part of the closure contract?  Do we do anything special if the count comes back as zero? Or is this a hint to the user on how to best to bail if they are also unwilling to throw an error.

Reply:
We do actually; it's possible to form a SmallString with 0 length that is not canonicalized to the empty SmallString, so we have to detect that.

Reply:
Hah, right. But, the caller is the one who is throwing away the result, right?

Reply:
Oops, missed this. Yes.